src/com/goide/highlighting/exitpoint/GoBreakStatementExitPointHandler.java

```diff
@@ -0,0 +1,149 @@
+ * limitations under the License.
+ */
+
+package com.goide.highlighting.exitpoint;
+
+import com.goide.psi.*;
+import com.intellij.codeInsight.highlighting.HighlightUsagesHandlerBase;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.psi.util.PsiTreeUtil;
+import com.intellij.util.Consumer;
+import com.intellij.util.containers.ContainerUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+
+public class GoBreakStatementExitPointHandler extends HighlightUsagesHandlerBase<PsiElement> {
+  @NotNull private final PsiElement myTarget;
+  @Nullable private final GoBreakStatement myBreakStatement;
+  @Nullable private PsiElement myOwner;
+
+  private GoBreakStatementExitPointHandler(Editor editor,
+                                           PsiFile file,
+                                           @NotNull PsiElement target,
+                                           @Nullable GoBreakStatement breakStatement,
+                                           @Nullable PsiElement owner) {
+    super(editor, file);
+    myTarget = target;
+    myBreakStatement = breakStatement;
+    myOwner = owner;
+  }
+
+  @NotNull
+  @Override
+  public List<PsiElement> getTargets() {
+    return ContainerUtil.newSmartList(myTarget);
+  }
+
+  @Override
+  protected void selectTargets(List<PsiElement> targets, @NotNull Consumer<List<PsiElement>> selectionConsumer) {
+    selectionConsumer.consume(targets);
+  }
+
+  @Override
+  public void computeUsages(List<PsiElement> targets) {
+    final PsiElement breakStmtOwner;
+    if (myOwner != null) {
+      breakStmtOwner = myOwner;
+    }
+    else if (myBreakStatement != null) {
+      breakStmtOwner = getBreakStatementOwner(myBreakStatement);
+    }
+    else {
+      breakStmtOwner = null;
+    }
+    GoRecursiveVisitor visitor = new GoRecursiveVisitor() {
+      @Override
+      public void visitLabelDefinition(@NotNull GoLabelDefinition o) {
+        if (o == breakStmtOwner) {
+          addOccurrence(o);
+        }
+        super.visitLabelDefinition(o);
+      }
+
+      @Override
+      public void visitBreakStatement(@NotNull GoBreakStatement o) {
+        if (o == myBreakStatement || getBreakStatementOwner(o) == breakStmtOwner) {
+          addOccurrence(o);
+        }
+        super.visitBreakStatement(o);
+      }
+
+      @Override
+      public void visitSwitchStatement(@NotNull GoSwitchStatement o) {
+        if (o == breakStmtOwner) {
+          GoSwitchStart switchStart = o.getSwitchStart();
+          if (switchStart != null) {
+            addOccurrence(switchStart.getSwitch());
+          }
+        }
+        super.visitSwitchStatement(o);
+      }
+
+      @Override
+      public void visitForStatement(@NotNull GoForStatement o) {
+        if (o == breakStmtOwner) {
+          addOccurrence(o.getFor());
```

Comment:
Okay, so here's my proposal for how break statements with labels should be highlighted:

1) Only break statements within the same function should be highlighted (which should fix the second screenshot you provided)
2) When caret is on the break statement with a label, we should highlight everything that it breaks out of: any `for`, `select`, and `switch` statement, as well as the label. However, my concern is that it may be confusing if the statements are far apart.
3) When caret is on `for`, `select`, or `switch`, we should not highlight any labelled break statements, because it is complicated to see whether the label they refer to is outside or inside the owner.

@zolotov what do you think?